InvalidAbsFile "/usr/lib64/ghc-9.2.3/lib/../lib/x86_64-linux-ghc-9.2.3/rts-1.0.2/include/ghcversion.h"

### General summary/comments (optional)

This is a bit annoying: afaict this is triggered by Stackage nightly or lts switching to a newer minor ghc release.

Note the file no longer exists (the system package was updated to 9.2.4).

I think the workaround is to run `stack --resolver nightly clean` in this case.

### Steps to reproduce

I haven't tried to verify how to reproduce exactly but I imagine:

1. Build my project with previous minor ghc Stackage snapshot
2. Try to rebuild with current minor ghc Stackage snapshot

In practice I am actually using `stack-all` for my building for nightly and lts.

### Expected
Expect stack to ignore old out of date (removed) minor ghc
and build freshly with current minor ghc.

### Actual

~~~text
stack-all -d
# nightly-2022-09-17
stack -v --resolver nightly-2022-09-17 build

[see below for package version output]

Warning: this is an unsupported build that may use different versions of
dependencies and GHC than the officially released binaries, and therefore may
not behave identically.  If you encounter problems, please try the latest
official build by running 'stack upgrade --force-download'.

2022-09-17 16:00:15.694322: [debug] Checking for project config at: /var/home/petersen/github/simple-cmd/stack.yaml
2022-09-17 16:00:15.694545: [debug] Loading project config file stack.yaml
2022-09-17 16:00:15.697218: [debug] Using resolver: nightly-2022-09-17 specified on command line
2022-09-17 16:00:15.697332: [debug] Using package location completions from a lock file
2022-09-17 16:00:15.817134: [debug] Prefetching git repos: []
2022-09-17 16:00:15.817206: [debug] []
2022-09-17 16:00:15.817960: [debug] Asking for a supported GHC version
2022-09-17 16:00:15.818155: [debug] Installed tools: 
 - ghc-tinfo6-8.0.2
 - ghc-tinfo6-8.10.7
 - ghc-tinfo6-8.2.2
 - ghc-tinfo6-8.4.4
 - ghc-tinfo6-8.6.5
 - ghc-tinfo6-8.8.4
 - ghc-tinfo6-9.0.2
 - ghc-tinfo6-9.2.2
 - ghc-tinfo6-7.10.3
 - ghc-tinfo6-9.2.3
 - ghc-tinfo6-9.2.4
 - ghc-tinfo6-9.4.2
 - ghc-tinfo6-9.4.1
2022-09-17 16:00:15.818683: [debug] Run process: /sbin/ldconfig -p
2022-09-17 16:00:15.819563: [debug] Process finished in 1ms: /sbin/ldconfig -p
2022-09-17 16:00:15.819988: [debug] Did not find shared library libtinfo.so.5
2022-09-17 16:00:15.820013: [debug] Found shared library libtinfo.so.6 in 'ldconfig -p' output
2022-09-17 16:00:15.820044: [debug] Found shared library libncursesw.so.6 in 'ldconfig -p' output
2022-09-17 16:00:15.820061: [debug] Found shared library libgmp.so.10 in 'ldconfig -p' output
2022-09-17 16:00:15.820094: [debug] Did not find shared library libgmp.so.3
2022-09-17 16:00:15.820106: [debug] Potential GHC builds: tinfo6, ncurses6
2022-09-17 16:00:15.820133: [debug] Found already installed GHC builds: tinfo6
2022-09-17 16:00:15.820234: [debug] (SQL) SELECT "id","actual_version","arch","ghc_path","ghc_size","ghc_modified","ghc_pkg_path","runghc_path","haddock_path","cabal_version","global_db","global_db_cache_size","global_db_cache_modified","info","global_dump" FROM "compiler_cache" WHERE "ghc_path"=?; [PersistText "/var/home/petersen/.stack/programs/x86_64-linux/ghc-tinfo6-9.2.4/bin/ghc"]
2022-09-17 16:00:15.844401: [debug] Loaded compiler information from cache
2022-09-17 16:00:15.844552: [debug] Asking for a supported GHC version
2022-09-17 16:00:15.844575: [debug] Resolving package entries
2022-09-17 16:00:15.844606: [debug] Parsing the targets
2022-09-17 16:00:15.847283: [debug] Checking flags
2022-09-17 16:00:15.847351: [debug] SourceMap constructed
2022-09-17 16:00:15.851082: [debug] Starting to execute command inside EnvConfig
2022-09-17 16:00:15.852835: [debug] Finding out which packages are already installed
2022-09-17 16:00:15.852938: [debug] Run process: /usr/bin/ghc-pkg-9.2.4 --global --no-user-package-db dump --expand-pkgroot
2022-09-17 16:00:15.881239: [debug] Process finished in 28ms: /usr/bin/ghc-pkg-9.2.4 --global --no-user-package-db dump --expand-pkgroot
2022-09-17 16:00:15.887157: [debug] Ignoring package zlib due to wanting version 0.6.3.0 instead of 0.6.2.3
2022-09-17 16:00:15.887318: [debug] Run process: /usr/bin/ghc-pkg-9.2.4 --user --no-user-package-db --package-db /var/home/petersen/.stack/snapshots/x86_64-linux-tinfo6/fa21c8874354f1328fd5f80b267aec038134cb4ae42e0b3fd9881d543666581f/9.2.4/pkgdb dump --expand-pkgroot
2022-09-17 16:00:15.905819: [debug] Process finished in 18ms: /usr/bin/ghc-pkg-9.2.4 --user --no-user-package-db --package-db /var/home/petersen/.stack/snapshots/x86_64-linux-tinfo6/fa21c8874354f1328fd5f80b267aec038134cb4ae42e0b3fd9881d543666581f/9.2.4/pkgdb dump --expand-pkgroot
2022-09-17 16:00:15.906114: [debug] Run process: /usr/bin/ghc-pkg-9.2.4 --user --no-user-package-db --package-db /var/home/petersen/github/simple-cmd/.stack-work/install/x86_64-linux-tinfo6/fa21c8874354f1328fd5f80b267aec038134cb4ae42e0b3fd9881d543666581f/9.2.4/pkgdb dump --expand-pkgroot
2022-09-17 16:00:15.921855: [debug] Process finished in 16ms: /usr/bin/ghc-pkg-9.2.4 --user --no-user-package-db --package-db /var/home/petersen/github/simple-cmd/.stack-work/install/x86_64-linux-tinfo6/fa21c8874354f1328fd5f80b267aec038134cb4ae42e0b3fd9881d543666581f/9.2.4/pkgdb dump --expand-pkgroot
2022-09-17 16:00:15.922184: [debug] Constructing the build plan
2022-09-17 16:00:15.923060: [debug] Checking if we are going to build multiple executables with the same name
2022-09-17 16:00:15.923115: [debug] Executing the build plan
2022-09-17 16:00:15.923945: [debug] (SQL) SELECT "id","directory","type","pkg_src","active","path_env_var","haddock" FROM "config_cache" WHERE "directory"=? AND "type"=?; [PersistText "/var/home/petersen/github/simple-cmd/",PersistText "config"]
2022-09-17 16:00:15.924171: [debug] (SQL) SELECT "id", "config_cache_id", "index", "option" FROM "config_cache_dir_option" WHERE ("config_cache_id"=?) ORDER BY "index"; [PersistInt64 1]
2022-09-17 16:00:15.924300: [debug] (SQL) SELECT "id", "config_cache_id", "index", "option" FROM "config_cache_no_dir_option" WHERE ("config_cache_id"=?) ORDER BY "index"; [PersistInt64 1]
2022-09-17 16:00:15.924395: [debug] (SQL) SELECT "id", "config_cache_id", "ghc_pkg_id" FROM "config_cache_dep" WHERE ("config_cache_id"=?); [PersistInt64 1]
2022-09-17 16:00:15.924466: [debug] (SQL) SELECT "id", "config_cache_id", "component" FROM "config_cache_component" WHERE ("config_cache_id"=?); [PersistInt64 1]
2022-09-17 16:00:15.924757: [debug] Start: getPackageFiles /var/home/petersen/github/simple-cmd/simple-cmd.cabal
2022-09-17 16:00:15.926318: [debug] Finished with exception in 2ms: getPackageFiles /var/home/petersen/github/simple-cmd/simple-cmd.cabal
Exception thrown: InvalidAbsFile "/usr/lib64/ghc-9.2.3/lib/../lib/x86_64-linux-ghc-9.2.3/rts-1.0.2/include/ghcversion.h"
2022-09-17 16:00:15.926550: [error] InvalidAbsFile "/usr/lib64/ghc-9.2.3/lib/../lib/x86_64-linux-ghc-9.2.3/rts-1.0.2/include/ghcversion.h"
stack-all: nightly build failed
~~~

### Stack version

~~~text
Version 2.9.0.1, Git revision a90609a07ff6fe85e5eb89a7c2317e3c9d0fe02b (dirty) (50 commits) RELEASE-CANDIDATE x86_64
~~~
built with `stack-2.7.5 --resolver nightly install`.

Same error occurs with stack-2.7.5.

Perhaps I am barking up the wrong tree. Looking at the errors reported:

~~~text
2022-09-17 16:00:15.924757: [debug] Start: getPackageFiles /var/home/petersen/github/simple-cmd/simple-cmd.cabal
2022-09-17 16:00:15.926318: [debug] Finished with exception in 2ms: getPackageFiles /var/home/petersen/github/simple-cmd/simple-cmd.cabal
Exception thrown: InvalidAbsFile "/usr/lib64/ghc-9.2.3/lib/../lib/x86_64-linux-ghc-9.2.3/rts-1.0.2/include/ghcversion.h"
2022-09-17 16:00:15.926550: [error] InvalidAbsFile "/usr/lib64/ghc-9.2.3/lib/../lib/x86_64-linux-ghc-9.2.3/rts-1.0.2/include/ghcversion.h"
stack-all: nightly build failed
~~~

~~~text
BNFC         > configure (lib + exe)
[1 of 2] Compiling Main             ( /Users/abel/project/open-source/bnfc/source/Setup.lhs, /Users/abel/project/open-source/bnfc/source/.stack-work/dist/x86_64-osx/Cabal-3.6.3.0/setup/Main.o )
[2 of 2] Compiling StackSetupShim   ( /Users/abel/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /Users/abel/project/open-source/bnfc/source/.stack-work/dist/x86_64-osx/Cabal-3.6.3.0/setup/StackSetupShim.o )
Linking /Users/abel/project/open-source/bnfc/source/.stack-work/dist/x86_64-osx/Cabal-3.6.3.0/setup/setup ...
Configuring BNFC-2.9.5...
Completed 5 action(s).
InvalidAbsFile "/usr/local/lib/ghc-9.2.2/lib/../lib/x86_64-osx-ghc-9.2.2/rts-1.0.2/include/ghcversion.h"
~~~

~~~text
2022-11-12T18:27:18.5311560Z [0magda2lagda> configure (exe)[0m
2022-11-12T18:27:19.5788730Z [0mConfiguring agda2lagda-0.2021.6.1...[0m
2022-11-12T18:27:20.1155100Z [0mInvalidAbsFile "/Users/runner/.ghcup/ghc/9.2.2/lib/ghc-9.2.2/lib/../lib/x86_64-osx-ghc-9.2.2/rts-1.0.2/include/ghcversion.h"[0m
2022-11-12T18:27:20.1551270Z ##[error]Process completed with exit code 1.
~~~

The problem is not that the file does not exist. The problem is that the path to the file given is not in the correct format for an absolute path to a file - it contains an `..` component, which makes it a relative file path. As explained in (`path` package) `Path.Posix.parseAbsFile`:
~~~haskell
-- | Convert an absolute 'FilePath' to a normalized absolute file 'Path'.
--
-- Throws: 'InvalidAbsFile' when the supplied path:
--
-- * is not an absolute path
-- * is a directory path i.e.
--
--     * has a trailing path separator
--     * is @.@ or ends in @/.@
--
-- * contains a @..@ path component representing the parent directory
-- * is not a valid path (See 'FilePath.isValid')
--
parseAbsFile :: MonadThrow m
             => FilePath -> m (Path Abs File)
parseAbsFile filepath =
  case validAbsFile filepath of
    True
      | normalized <- normalizeFilePath filepath
      , validAbsFile normalized ->
        return (Path normalized)
    _ -> throwM (InvalidAbsFile filepath)
~~~

--- a/src/Path/Extra.hs
+++ b/src/Path/Extra.hs
@@ -14,21 +14,27 @@ module Path.Extra
   , pathToLazyByteString
   , pathToText
   , tryGetModificationTime
+  , forgivingResolveFile
+  , forgivingResolveFile'
   ) where
 
 import           Data.Time ( UTCTime )
 import           Path
-                   ( Abs, Dir, File, Rel, parseAbsDir, parseAbsFile
-                   , toFilePath
+                   ( Abs, Dir, File, PathException (..), Rel, parseAbsDir
+                   , parseAbsFile, toFilePath
                    )
 import           Path.Internal ( Path (Path) )
-import           Path.IO ( doesDirExist, doesFileExist, getModificationTime )
+import           Path.IO
+                   ( doesDirExist, doesFileExist, getCurrentDir
+                   , getModificationTime
+                   )
 import           RIO
 import           System.IO.Error ( isDoesNotExistError )
 import qualified Data.ByteString.Char8 as BS
 import qualified Data.ByteString.Lazy.Char8 as BSL
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
+import qualified System.Directory as D
 import qualified System.FilePath as FP
 
 -- | Convert to FilePath but don't add a trailing slash.
@@ -123,3 +129,30 @@ pathToText = T.pack . toFilePath
 
 tryGetModificationTime :: MonadIO m => Path Abs File -> m (Either () UTCTime)
 tryGetModificationTime = liftIO . tryJust (guard . isDoesNotExistError) . getModificationTime
+
+-- | 'Path.IO.resolveFile' (@path-io@ package) throws 'InvalidAbsFile' (@path@
+-- package) if the file does not exist; this function yields 'Nothing'.
+forgivingResolveFile ::
+     MonadIO m
+  => Path Abs Dir
+     -- ^ Base directory
+  -> FilePath
+     -- ^ Path to resolve
+  -> m (Maybe (Path Abs File))
+forgivingResolveFile b p = liftIO $
+  D.canonicalizePath (toFilePath b FP.</> p) >>= \cp ->
+    catch
+      (Just <$> parseAbsFile cp)
+      ( \e -> case e of
+          InvalidAbsFile _ -> pure Nothing
+          _ -> throwIO e
+      )
+
+-- | 'Path.IO.resolveFile'' (@path-io@ package) throws 'InvalidAbsFile' (@path@
+-- package) if the file does not exist; this function yields 'Nothing'.
+forgivingResolveFile' ::
+     MonadIO m
+  => FilePath
+     -- ^ Path to resolve
+  -> m (Maybe (Path Abs File))
+forgivingResolveFile' p = getCurrentDir >>= flip forgivingResolveFile p

--- a/src/Stack/Build/Execute.hs
+++ b/src/Stack/Build/Execute.hs
@@ -76,11 +76,14 @@ import           Path
                    , stripProperPrefix
                    )
 import           Path.CheckInstall ( warnInstallSearchPathIssues )
-import           Path.Extra ( toFilePathNoTrailingSep, rejectMissingFile )
+import           Path.Extra
+                   ( forgivingResolveFile, rejectMissingFile
+                   , toFilePathNoTrailingSep
+                   )
 import           Path.IO
                    ( copyFile, doesDirExist, doesFileExist, ensureDir
-                   , forgivingAbsence, ignoringAbsence, removeDirRecur
-                   , removeFile, renameDir, renameFile, resolveFile
+                   , ignoringAbsence, removeDirRecur, removeFile, renameDir
+                   , renameFile
                    )
 import           RIO.Process
                    ( HasProcessContext, byteStringInput, doesExecutableExist
@@ -660,7 +663,7 @@ copyExecutables exes = do
             case loc of
                 Snap -> snapBin
                 Local -> localBin
-    mfp <- liftIO $ forgivingAbsence (resolveFile bindir $ T.unpack name ++ ext)
+    mfp <- liftIO $ forgivingResolveFile bindir (T.unpack name ++ ext)
       >>= rejectMissingFile
     case mfp of
       Nothing -> do
@@ -2541,7 +2544,7 @@ mungeBuildOutput excludeTHLoading makeAbsolute pkgDir compilerVer = void $
       if isValidSuffix y
         then liftIO $
           fmap (fmap ((T.takeWhile isSpace x <>) . T.pack . toFilePath)) $
-            forgivingAbsence (resolveFile pkgDir (T.unpack $ T.dropWhile isSpace x)) `catch`
+            forgivingResolveFile pkgDir (T.unpack $ T.dropWhile isSpace x) `catch`
               \(_ :: PathException) -> pure Nothing
         else pure Nothing
     case mabs of

--- a/src/Stack/ComponentFile.hs
+++ b/src/Stack/ComponentFile.hs
@@ -43,11 +43,12 @@ import           Path
                    , stripProperPrefix
                    )
 import           Path.Extra
-                   ( parseCollapsedAbsFile, rejectMissingDir, rejectMissingFile
+                   ( forgivingResolveFile, parseCollapsedAbsFile
+                   , rejectMissingDir, rejectMissingFile
                    )
 import           Path.IO
                    ( doesDirExist, doesFileExist, forgivingAbsence
-                   , getCurrentDir, listDir, resolveDir, resolveFile
+                   , getCurrentDir, listDir, resolveDir
                    )
 import           Stack.Constants
                    ( haskellDefaultPreprocessorExts, haskellFileExts
@@ -294,7 +295,7 @@ parseHI hiPath = do
                         Iface.unList . Iface.dmods . Iface.deps
           resolveFileDependency file = do
             resolved <-
-              liftIO (forgivingAbsence (resolveFile dir file)) >>=
+              liftIO (forgivingResolveFile dir file) >>=
                 rejectMissingFile
             when (isNothing resolved) $
               prettyWarnL

--- a/src/Stack/Ghci.hs
+++ b/src/Stack/Ghci.hs
@@ -30,7 +30,7 @@ import qualified Data.Text.Lazy as TL
 import qualified Data.Text.Lazy.Encoding as TLE
 import qualified Distribution.PackageDescription as C
 import           Path
-import           Path.Extra ( toFilePathNoTrailingSep )
+import           Path.Extra ( forgivingResolveFile', toFilePathNoTrailingSep )
 import           Path.IO hiding ( withSystemTempDir )
 import           RIO.Process
                    ( HasProcessContext, exec, proc, readProcess_
@@ -239,7 +239,7 @@ preprocessTargets buildOptsCLI sma rawTargets = do
         then do
             fileTargets <- forM fileTargetsRaw $ \fp0 -> do
                 let fp = T.unpack fp0
-                mpath <- liftIO $ forgivingAbsence (resolveFile' fp)
+                mpath <- liftIO $ forgivingResolveFile' fp
                 case mpath of
                     Nothing -> throwM (MissingFileTarget fp)
                     Just path -> pure path

--- a/src/Stack/PackageFile.hs
+++ b/src/Stack/PackageFile.hs
@@ -16,8 +16,7 @@ import           Distribution.PackageDescription hiding ( FlagName )
 import           Distribution.Simple.Glob ( matchDirFileGlob )
 import qualified Distribution.Types.UnqualComponentName as Cabal
 import           Path ( parent )
-import           Path.Extra ( rejectMissingFile )
-import           Path.IO ( forgivingAbsence, resolveFile )
+import           Path.Extra ( forgivingResolveFile, rejectMissingFile )
 import           Stack.ComponentFile
                    ( benchmarkFiles, executableFiles, libraryFiles
                    , resolveOrWarn, testFiles
@@ -37,7 +36,7 @@ resolveFileOrWarn :: FilePath.FilePath
                   -> RIO GetPackageFileContext (Maybe (Path Abs File))
 resolveFileOrWarn = resolveOrWarn "File" f
  where
-  f p x = liftIO (forgivingAbsence (resolveFile p x)) >>= rejectMissingFile
+  f p x = liftIO (forgivingResolveFile p x) >>= rejectMissingFile
 
 -- | Get all files referenced by the package.
 packageDescModulesAndFiles ::

